### PR TITLE
feat(chart): Update linting and checking for apiVersion v2

### PR DESCRIPTION
**Pull request by hickeyma**
Update linting and checking for apiVersion v1 and v2.

Some work previously done like in PR #5730.

Partial close for: #5907

**What this PR does / why we need it**:
Helm v3 has introduced new constructs in the charts for apiVersion v2. Checks needed to handle the changes between apiVersion v1 and v2.

**Special notes for your reviewer**:
What changes to annotations were introduced in v3?

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

File: pkg/chart/loader/load.go
Comment (bacongobbler) [inline pkg/chart/loader/load.go]: requirements.yaml is deprecated, but we will continue to support it in v2. We could probably inform the user that they should migrate to using Chart.yaml instead but we still should support this if they bump versions. See line 104 :)

**Review by bacongobbler — CHANGES_REQUESTED** [review]

File: pkg/chart/loader/load.go
Comment (hickeyma) [reply pkg/chart/loader/load.go]: Good point @bacongobbler. Someone may bump it but maintain the requirements file. Updated accordingly.

**Comment by hickeyma** [comment]
Ready for review again.

**Review by bacongobbler — APPROVED** [review]

File: pkg/chart/loader/load.go
Comment (mattfarina) [inline pkg/chart/loader/load.go]: This is a warning if the `apiVersion` is `v2` and there is a `requirements.yaml` file. Would it be too much to remove the `requirements.yaml` from the `v2` chart and make this an error?

File: pkg/chart/loader/testdata/frobnitz.v2.reqs/Chart.lock
Comment (mattfarina) [inline pkg/chart/loader/testdata/frobnitz.v2.reqs/Chart.lock]: If there are dependencies in the lock file shouldn't the appropriate piece be in the `Chart.yaml` file, too?

File: pkg/chart/loader/load.go
Comment (hickeyma) [reply pkg/chart/loader/load.go]: @mattfarina There was feed back from @bacongobbler in https://github.com/helm/helm/pull/6117#discussion_r308900043 about handling the requirements file and giving a warning.

File: pkg/chart/loader/testdata/frobnitz.v2.reqs/Chart.lock
Comment (hickeyma) [reply pkg/chart/loader/testdata/frobnitz.v2.reqs/Chart.lock]: That could only be described as a human error mistake! :) Apologies, will remove.

File: pkg/chart/loader/load.go
Comment (bacongobbler) [reply pkg/chart/loader/load.go]: Because the translation from requirements.yaml to Chart.yaml are 1:1 there's no technical reason why we can't support both in apiVersion v2.

Comment (mattfarina) [reply pkg/chart/loader/load.go]: @bacongobbler @adamreese @hickeyma Are we thinking of these as specs we can independently document and have multiple tools work on things that conform to the specs? With the `v1` we didn't do that even though there was a task for it. If we do think of it as a spec... than what does the spec say about the location of the requirements? If one of us writes the spec on this how do we describe the differences?

Comment (bacongobbler) [reply pkg/chart/loader/load.go]: I'll defer to @adamreese, but you do raise good points RE: third-party tools looking to support apiVersion v2.

**Review by bacongobbler — CHANGES_REQUESTED** [review]
rescinding LGTM due to ongoing conversations

File: pkg/chart/loader/load.go
Comment (bacongobbler) [reply pkg/chart/loader/load.go]: @mattfarina would it be fair to move forward with this PR as-is and carry on the discussion in a ticket? This retains the same functionality as before. We can discuss whether or not we want to not support requirements.yaml in apiVersion v2 as a separate PR. How does that sound?

Comment (mattfarina) [reply pkg/chart/loader/load.go]: That sounds fine. I started #6151 to have that conversation.

Note, the whole reason I thought of a spec being written out is a) I've been asked about one and b) it would cause us to think some things through that we may be missing.

File: pkg/chart/loader/testdata/frobnitz.v2.reqs/Chart.lock
Comment (mattfarina) [reply pkg/chart/loader/testdata/frobnitz.v2.reqs/Chart.lock]: If you add the dependencies to the `Chart.yaml` I'll come through and approve this one given the other conversation is moving to a follow-up.

File: pkg/chart/loader/load.go
Comment (hickeyma) [reply pkg/chart/loader/load.go]: @mattfarina Good idea. Putting it out in the open will help us and the users afterwards.

File: pkg/chart/loader/testdata/frobnitz.v2.reqs/Chart.lock
Comment (hickeyma) [reply pkg/chart/loader/testdata/frobnitz.v2.reqs/Chart.lock]: The lock file should not have been there as the dependencies are in the requirements. The test is to show that v3 will handle chart specified as apiVersion "v2" but using dependencies in requirements file.

**Comment by hickeyma** [comment]
@bacongobbler @mattfarina ready for review again.

**Review by bacongobbler — APPROVED** [review]

**Comment by bacongobbler** [comment]
feel free to merge when ready

**Comment by hickeyma** [comment]
Thanks @bacongobbler. I'll wait on @mattfarina before I merge.

**Review by mattfarina — APPROVED** [review]

**Comment by hickeyma** [comment]
Thanks @mattfarina.

**Pull request merged**
